Add hashing primitives

Geod24 opened the pull request:
```
Those hashing primitives use `BLAKE2b` for hashing.
They allow struct to have control over the way hashing is done.
The results are currently not tested against other implementation,
but they are stable and will unblock the team.
```

We need a few fixes in `libsodiumd` for attributes to fully work, but that should unblock everyone.

Geod24 committed:
Add hashing primitives

Those hashing primitives use `BLAKE2b` for hashing.
They allow struct to have control over the way hashing is done.
The results are currently not tested against other implementation,
but they are stable and will unblock the team.

AndrejMitrovic commented on source/agora/common/Hash.d:
```diff
@@ -0,0 +1,159 @@
+
+
+///
+nothrow @nogc @safe unittest
```
It's a little weird to put the unittest first and then the implementation?

Geod24 replied:
Module documentation

AndrejMitrovic commented on source/agora/common/Hash.d:
```diff
@@ -0,0 +1,159 @@
+}
+
+/// Ditto
+public void hashPart (T) (scope const auto ref T record, scope HashDg state)
```
Shouldn't these be private?

AndrejMitrovic replied:
Ah no, I get it now. It's called by the struct's implementation of `computeHash`.

AndrejMitrovic commented on source/agora/common/Hash.d:
```diff
@@ -0,0 +1,159 @@
+    crypto_generichash_state state;
+    crypto_generichash_init(&state, null, 0, Hash.sizeof);
+    scope HashDg dg = (scope const(ubyte)[] data) @trusted
+                     => cast(void)crypto_generichash_update(&state, data.ptr, data.length);
```
What does `cast(void)` do?

Geod24 replied:
Avoid DMD erroring on "Cannot return non-`void` from `void` function"...

AndrejMitrovic replied:
I'm guessing it discards the return value, to help in type inference for the declaration of `dg`.

Geod24 replied:
That's a more accurate way to put it, yes :)

AndrejMitrovic commented:
Can you just move the unittests after the functions? The first thing I want to see is the code when I open the module, not the tests.

Other than that it LGTM and I think it's ready to go.

Geod24 commented:
The very first `unittest` should be documentation. I agree it's not good doc. Will rework.

AndrejMitrovic commented:
> The very first unittest should be documentation. I agree it's not good doc. Will rework.

But that's the point of `///` unittests, they document the symbol that's defined before the unittest.

Geod24 committed:
Make Hash type be 512 bits / 64 bytes

Since we are going to use BLAKE2b for the time being,
use the maximum possible size for BLAKE2b: 512 bits.
Note that using 256 bits should not create problems
security wise but is less common and less likely to
be supported by external implementations.

AndrejMitrovic commented:
So the documentation looks good with the unittest in place.

AndrejMitrovic commented:
@MukeunKim @TrustHenry want to review this?

AndrejMitrovic commented on source/agora/common/Hash.d:
```diff
@@ -0,0 +1,200 @@
+}
+
+/// Ditto
+public void hashPart () (scope const auto ref Hash record, scope HashDg state)
```
What's the extra `()` for?

Geod24 replied:
DMD complains otherwise (`auto ref`)
